### PR TITLE
Loki: Make label browser accessible in query builder

### LINKED ISSUE #57744: Loki: Make label browser accessible from both code and builder modes
### Summary

Label browser is powerful and praised feature that helps to search trough labels, that is currently accessible and discoverable only trough code mode. To improve accessibility and discoverability, it should be accessible in both - code and builder modes.

### Problem statement

Label browser is very valuable feature that lets you easily navigate through your list of labels and filter and search trough values.  This feature is valuable for users using both - code and builder editor mode, therefore label browser should be accessible from both modes. Moreover, it is currently not discoverable for new users who are using only builder mode.

During user interviews we learned that some users switch between builder-code modes just to use label browser and then go back to builder mode, which is unnecessary. Moreover, this very useful feature is not discoverable for new users who are using only builder mode.

We already have other interactions that work in both modes, so there is not a reason why label browser should not be in builder mode. 

### Scope

Implementation (coding task)

### Solution

In query builder header, we already have a place for features/interactions that are present in both modes (e.g. Kick start your query). We should move label browser button to query builder header and make it accessible in all query modes. 

Moreover, we should evaluate and test if modal wouldn't be a better solution for label browser, as this is what we use in query patterns and would be cohesive. It would also make it much easier to use in dashboard.

### PR DESCRIPTION
**Pull request by itsgareth**
**What is this feature?**
This PR makes the Label browser accessible from both the query builder and the code mode

**Why do we need this feature?**
During user interviews we learned that some users switch from the query builder to the code mode just to use label browser and then go back to builder mode, which is unnecessary

**Who is this feature for?**
...

**Which issue(s) does this PR fix?**:
Fixes #57744

**Special notes for your reviewer**:
...

**Before:**
<div>
Screenshot 2022-11-21 at 15 43 47
Screenshot 2022-11-21 at 15 45 47
</div>

**After:**
<div>
Screenshot 2022-11-21 at 15 43 21
Screenshot 2022-11-21 at 15 45 24
</div>

### CODE REVIEW

**Review by matyax — CHANGES_REQUESTED** [review]
Good work! There are some minor things to adjust.

File: public/app/plugins/datasource/loki/querybuilder/components/LabelBrowserModal.test.tsx [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+  it('renders the label browser modal when open', () => {
+    render(<LabelBrowserModal {...props} />);
+    expect(screen.getByRole('heading', { name: /label browser/i })).toBeInTheDocument();
+  });
+
+  it("doesn't render the label browser modal when closed", () => {
+    render(<LabelBrowserModal {...props} isOpen={false} />);
+    expect(screen.queryByRole('heading', { name: /label browser/i })).toBeNull();
+  });
```
Comment (matyax) [inline public/app/plugins/datasource/loki/querybuilder/components/LabelBrowserModal.test.tsx:41]: Would also love to have a test case for `onChange` when we change the query. Feel free to drop it in another PR.

**Review by matyax — COMMENTED** [review]
Looks good! One minor adjustment and it's done.

File: public/app/plugins/datasource/loki/querybuilder/components/LabelBrowserModal.tsx [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+export const LabelBrowserModal = (props: Props) => {
+  const { isOpen, onClose, languageProvider, app } = props;
+
+  const LAST_USED_LABELS_KEY = 'grafana.datasources.loki.browser.labels';
+
+  const changeQuery = (value: string, override?: boolean) => {
+    const { query, onChange, onRunQuery } = props;
+
+    if (onChange) {
```
[the remark this answers is not in the document]
Comment (gwdawson) [reply public/app/plugins/datasource/loki/querybuilder/components/LabelBrowserModal.tsx:29]: nice catch 😎

**Review by matyax — COMMENTED** [review]
One last change. Sorry I didn't see it before!

File: public/app/plugins/datasource/loki/components/LokiLabelBrowser.tsx [hunk 3]
```diff
@@ -467,7 +466,12 @@ export class UnthemedLokiLabelBrowser extends React.Component<BrowserProps, Brow
             2. Find values for the selected labels
           </Label>
           <div>
-            <Input onChange={this.onChangeSearch} aria-label="Filter expression for values" value={searchTerm} />
+            <Input
+              onChange={this.onChangeSearch}
+              aria-label="Filter expression for values"
+              value={searchTerm}
+              placeholder={'Enter a label name'}
```
Comment (matyax) [inline public/app/plugins/datasource/loki/components/LokiLabelBrowser.tsx:473]: I misinterpreted the explanation of this field, so `Enter a label value` was correct. Feel free to apply the suggestion from here.

```suggestion
              placeholder={'Enter a label value'}
```

imagen

File: public/app/plugins/datasource/loki/components/LokiLabelBrowser.tsx [hunk 4]
```diff
@@ -467,7 +466,12 @@ export class UnthemedLokiLabelBrowser extends React.Component<BrowserProps, Brow
...
+            />
           </div>
           <div className={styles.valueListArea}>
             {selectedLabels.map((label) => (
```
Comment (matyax) [inline public/app/plugins/datasource/loki/components/LokiLabelBrowser.tsx:477]: While testing, my Loki instance went offline, so I had pre-fetched labels, but then it failed to make further requests.

With that, I saw that we display 2 messages that now are broken:

Message 1
Message 2

We need to adjust that.

To reproduce it, load the page with your Loki docker image on, and then turn it off to interact with the component.

I think moving it above the buttons would be acceptable.

imagen

**Review by matyax — COMMENTED** [review]
imagen

🙏

File: public/app/plugins/datasource/loki/components/LokiLabelBrowser.tsx (hunk 4 above)
Comment (matyax) [reply public/app/plugins/datasource/loki/components/LokiLabelBrowser.tsx:477]: Also I don't know what does "facetting" mean in that context at all.

Comment (matyax) [reply public/app/plugins/datasource/loki/components/LokiLabelBrowser.tsx:477]: After https://raintank-corp.slack.com/archives/C03PMJLVC67/p1669214990835099 I'd suggest rename "Facetting" with "Loading" please. Thanks!

**Review by matyax — APPROVED** [review]
GJ!

File: public/app/plugins/datasource/loki/querybuilder/components/LabelBrowserModal.test.tsx (hunk 1 above)
Comment (matyax) [reply public/app/plugins/datasource/loki/querybuilder/components/LabelBrowserModal.test.tsx:41]: This would be a good candidate for a follow-up task.

**Review by ivanahuckova — COMMENTED** [review]
I left couple of feedback bellow. Also I remember that we talked about making it wider - slightly smaller than full screen, but taking advantage of the full screen size. Is there a reason you decided not to make it larger, now when we have modal and space?

File: public/app/plugins/datasource/loki/components/LokiQueryEditor.tsx [hunk 5]
```diff
@@ -70,6 +73,30 @@ export const LokiQueryEditor = React.memo<LokiQueryEditorProps>((props) => {
     onChange(query);
   };
 
+  const onClickChooserButton = () => {
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/components/LokiQueryEditor.tsx:76]: `onClickChooserButton` -> `onClickLabelBrowserButton`

**Comment by gwdawson** [comment]
@ivanahuckova i like these suggestions! i'll make a pr for these changes.

> Also I remember that we talked about making it wider - slightly smaller than full screen, but taking advantage of the full screen size. Is there a reason you decided not to make it larger

https://github.com/grafana/grafana/pull/58525#discussion_r1028306646 - I wasn't sure how this will effect any of the other modals that already exist. my plan was to take a look to see if theres a way to increase the size for only a specific modal, in this case the label browser. if you think i would be better to just increase the size for all of them please let me know.

**Pull request merged**